### ISSUE TITLE
Openshift v3 deployed on OpenStack with Alcatel-Lucent Nuage SDN solution

### ISSUE THREAD

**Issue by roldancer** [openshift/origin#1919]
Hi All,
     I'm planning to deploy Openshift V3 beta on OpenStack (Redhat), this corporate IaaS  uses Nuage SDN solution for OpenStack networking, I would like to know if you think that this deployment could be problematic I saying this because we are going to have a SDN (OpenShift) inside another SDN (Nuage), any recommendation about this topology ?
I would like to hear you experience and recommendations ....

Many thanks.

**Comment by smarterclayton** [comment]
The expectation would be that in the presence of another SDN you would customize OpenShift to use that solution (via the networkplugin mechanism on the nodes) rather than using OpenShift SDN.  The hooks on each node allow you to intercept the pod IP assignment step and call out to a specific solution, which would be custom to the Nuage SDN.  If you need to do setup work before an individual node can request an IP, then you'll tend to implement that by watching for changes to the node topology from a client and updating Nuage with the relevant info.

@rajatchopra can you link in the relevant information @roldancer will need (we need to get this into the documentation eventually).

**Comment by roldancer** [comment]
Hi @smarterclayton any information/documentation will be welcomed, we will be working on this topology   next week with Redhat's pre-sale specialists, I think we can work together (Community/Redhat/we) in order to document this use case, at the enterprise level corporate SDN solution will be very common so sharing our experience will be useful for all.
Thanks.

**Comment by rajatchopra** [comment]
@roldancer Documentation is still in the pend, but hooks are already present for integration. The integration will typically happen from two ends. The node(s) and the master. 

* At the master end, there are watch APIs provided by OpenShift where your plugin code can watch new hosts being added to the cluster, new projects being created and new pods getting scheduled for cluster nodes.

* At the node end, call-out hooks exist. See how to organize the networking plugin for the nodes here:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/network/exec/exec.go#L17
Once installed, openshift can be called with --network-plugin='nuage_openshift_network' on nodes and all the networking is offloaded to the plugin in three steps:
  1. Init - plugin is called once everytime the node starts
  2. Setup - plugin is called each time a pod is created
  3. Teardown - plugin is called each time a pod is destroyed


This is exactly how openshift-sdn will also get hooked (though we will have a somewhat tighter integration as we re-use our code).

**Comment by FlorianOtel** [comment]
Hello all, 
(Full disclosure:  While I'm logged in here with my private GitHub credentials I'm writing this as NuageNetworks person responsible for this integration for Nuage customers in EMEA) 

Funny how the same issue gets circulated via several email threads (both private and corporate) and now GitHub...:)

We (NuageNetworks) are actively working on this: At the moment I'm typing this I'm installing Origin M4 in AWS  -- or at least trying to :) --   to better scope the LOE for any integration work that needs to happen on both sides (Nuage VSP and, respectively, OpenShift / Origin). 

While we have an initial estimate based on the respective architectures, a _proper_ scoping can only be done only after we put together a setup of Nuage VSP + OpenShift / Origin and assess that I place.

As I mentioned above, the former is already done, the latter is on-going as I type this. 

I would be very grateful for any guidance and help with OpenShift / Origin and work with any parties interested in this work / setup. 

Please connect either  to me -- or here -- for any further questions / comments. 

Thanks,  

@FlorianOtel

**Comment by FlorianOtel** [comment]
@roldancer the comment by @smarterclayton is spot-on.  Again, speaking solely from an initial evaluation pov (i.e. architecture based): 

Given that both OpenShift SDN and Nuage VSP are using OVS-based overlays (see http://www.slideshare.net/enakai/openshift-45465283 , slide 3) our _expectation_  (emphasis mine) is that the integration will happen at that level. Any deeper integration  (APIs, hooks) still TBD. 

Also, it seems this will depend significantly on the deployment environment i.e. the platform underlying OpenShift and the networking capabilities there... 

Like I mentioned in previous comment, I would be happy to work with any parties interesting in this.

**Comment by roldancer** [comment]
Hi @FlorianOtel, we would like to contribute to this issue, please let me know how can we work together.

Many thanks.

**Comment by rajatchopra** [comment]
Closing this issue as a specific Nuage plugin is being built using the hooks mentioned above.
